src/sources/kafka.rs

```diff
@@ -0,0 +1,331 @@
+    auto_offset_reset: String,
+    #[serde(default = "default_session_timeout_ms")]
+    session_timeout_ms: u64,
+    host_key: Option<String>,
+    key_field: Option<String>,
+}
+
+fn default_session_timeout_ms() -> u64 {
+    10000 // default in librdkafka
+}
+
+fn default_auto_offset_reset() -> String {
+    "largest".into() // default in librdkafka
+}
+
+#[typetag::serde(name = "kafka")]
+impl SourceConfig for KafkaSourceConfig {
+    fn build(
+        &self,
+        _name: &str,
+        _globals: &GlobalOptions,
+        out: mpsc::Sender<Event>,
+    ) -> Result<super::Source, String> {
+        kafka_source(self.clone(), out)
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+}
+
+fn kafka_source(
+    config: KafkaSourceConfig,
+    out: mpsc::Sender<Event>,
+) -> Result<super::Source, String> {
+    let consumer = Arc::new(create_consumer(config.clone())?);
+    let source = future::lazy(move || {
+        let hostname = hostname::get_hostname();
+        let host_key = config.host_key.clone().unwrap_or(event::HOST.to_string());
+        let consumer_ref = Arc::clone(&consumer);
+
+        // See https://github.com/fede1024/rust-rdkafka/issues/85#issuecomment-439141656
+        let stream = OwnedConsumerStream {
+            upstream: OwningHandle::new_with_fn(consumer, |c| {
+                let cf = unsafe { &*c };
+                Box::new(cf.start())
+            }),
+        };
+
+        stream
+            .then(move |message| {
+                match message {
+                    Err(e) => Err(error!("Error reading message from Kafka: {:?}", e)),
+                    Ok(Err(e)) => Err(error!("Kafka returned error: {:?}", e)),
+                    Ok(Ok(msg)) => {
+                        let payload = match msg.payload_view::<[u8]>() {
+                            None => return Err(()), // skip messages with empty payload
+                            Some(Err(e)) => return Err(error!("Cannot extract payload: {:?}", e)),
+                            Some(Ok(payload)) => Bytes::from(payload),
+                        };
+
+                        let message_key = if config.key_field.is_some() {
+                            match msg.key_view::<[u8]>() {
+                                None => None,
+                                Some(Err(e)) => return Err(error!("Cannot extract key: {:?}", e)),
+                                Some(Ok(key)) => Some(Bytes::from(key)),
+                            }
+                        } else {
+                            None
+                        };
+
+                        let event = create_event(
+                            payload,
+                            &config.key_field,
+                            message_key,
+                            &host_key,
+                            &hostname,
+                        );
+                        consumer_ref
+                            .store_offset(&msg)
```

Comment:
Is there a reason you prefer this to enabling auto commit? It'd be nice if we had a way to ack once an event is stored on disk or acked by a sink, but without that I'm not sure we get much benefit from manually committing offsets.

Reply:
Yes, [it is advised in `rust-rdkafka` docs](https://docs.rs/rdkafka/0.21.0/rdkafka/consumer/stream_consumer/struct.StreamConsumer.html) to disable auto commit and commit offsets manually, because otherwise some messages might be lost on stoping or restart. If there is a concern about performance of committing the offset after each message, the logic that would commit offsets less frequently should be implemented manually in Rust code.

Reply:
:+1: Sounds reasonable to me